src/types.rs

```diff
@@ -490,69 +494,397 @@ impl StreamMetadataBuilder {
     /// intermediaries to handle polls (say 10 seconds).
-    pub cache_control: Option<Duration>,
+    #[serde(
+        rename = "$cacheControl",
+        skip_serializing_if = "Option::is_none",
+        default
+    )]
+    pub cache_control: Option<u64>,
 
     /// The access control list for the stream.
-    pub acl: StreamAcl,
+    #[serde(
+        rename = "$acl",
+        skip_serializing_if = "Option::is_none",
+        deserialize_with = "deserialize_acl",
+        serialize_with = "serialize_acl",
+        default
+    )]
+    pub acl: Option<Acl>,
 
     /// An enumerable of key-value pairs of keys to JSON value for
     /// user-provided metadata.
+    #[serde(flatten, skip_serializing_if = "HashMap::is_empty", default)]
     pub custom_properties: HashMap<String, serde_json::Value>,
 }
 
+fn serialize_acl<S>(src: &Option<Acl>, serializer: S) -> std::result::Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    if let Some(ref acl) = src.as_ref() {
+        match acl {
+            Acl::User => serializer.serialize_str("$userStreamAcl"),
+            Acl::System => serializer.serialize_str("$systemStreamAcl"),
+            Acl::Stream(acl) => serializer.serialize_some(acl),
+        }
+    } else {
+        serializer.serialize_none()
+    }
+}
+
+fn deserialize_acl<'de, D>(deserializer: D) -> std::result::Result<Option<Acl>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    deserializer.deserialize_any(AclVisitor)
+}
+
+struct AclVisitor;
+
+impl<'de> Visitor<'de> for AclVisitor {
+    type Value = Option<Acl>;
+
+    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(formatter, "a EventStoreDB ACL")
+    }
+
+    fn visit_none<E>(self) -> std::result::Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        Ok(None)
+    }
+
+    fn visit_str<E>(self, value: &str) -> std::result::Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        match value {
+            "$userStreamAcl" => Ok(Some(Acl::User)),
+            "$systemStreamAcl" => Ok(Some(Acl::System)),
+            unknown => Err(serde::de::Error::invalid_value(
+                serde::de::Unexpected::Str(unknown),
+                &self,
+            )),
+        }
+    }
+
+    fn visit_map<M>(self, map: M) -> std::result::Result<Self::Value, M::Error>
+    where
+        M: serde::de::MapAccess<'de>,
+    {
+        let stream_acl =
+            Deserialize::deserialize(serde::de::value::MapAccessDeserializer::new(map))?;
+
+        Ok(Some(Acl::Stream(stream_acl)))
+    }
+}
+
 impl StreamMetadata {
+    pub fn new() -> Self {
+        StreamMetadata::default()
+    }
+
     /// Initializes a fresh stream metadata builder.
     pub fn builder() -> StreamMetadataBuilder {
         StreamMetadataBuilder::new()
     }
 }
 
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum Acl {
+    User,
+    System,
+    Stream(StreamAcl),
```

Comment:
@oskardudycz I'd like your input on how we could name those predefined ACLs, `User` (`$userStreamAcl` internally) and `System` (`$systemStreamAcl` internally)

Reply:
Done